Move settings registration to info.xml + merge 'personal' & 'admin' settings

tomneedham opened the pull request:
<!--
Thanks for submitting a change to ownCloud!

This is the bug tracker for the Server component. Find other components at https://github.com/owncloud/core/blob/master/.github/CONTRIBUTING.md#guidelines

For fixing potential security issues please see https://owncloud.org/security/

To make it possible for us to get your change reviewed and merged please fill out below information carefully.

Please note that any kind of change first has to be submitted to the master branch which holds the next major version of ownCloud.

We will carefully discuss if your change can or has to be backported to stable branches.
-->
## Description

Remove the legacy `OC_App::registerAdmin()` and `OC_App::registerPersonal()` calls in favour of a info.xml based approach. Apps will be able to register 'personal' and 'admin' settings panels by specifying class names in an XML element in the info.xml.

This also helps up break down our existing personal and admin settings page controllers into smaller chunks which can be tested easier.

Also the personal and settings pages have been combined (visually) although the routing remains the same.
## Related Issue

https://github.com/owncloud/core/issues/24188
## Motivation and Context

See https://github.com/owncloud/core/issues/24188
## How Has This Been Tested?
 - Manual testing
 - Unit tests on SettingManager and SettingsPageController
 - https://github.com/owncloud/QA/issues/335#issuecomment-277228055
## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->

<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

WIP (continuing w/c 24/10/16):
- [x] Fix registering of IPanels for core, and for Apps ⛏ 
- [x] Decide on category names
- [x] Add tests for SettingsManager ⚙ 
- [x] Add in catch for legacy calls to `OC_App:registerPersonal|Admin()`
- [x] Port over existing settings panels to IPanel and ISection ⏩ 
- [x] Write documentation updates ✏️

Commit:
WIP to loading settings panels from info.xml

PVince81 commented on lib/private/Settings/SettingsManager.php:
```diff
@@ -0,0 +1,276 @@
+      return [];
+    }
+    foreach($panels as $panelClassName) {
+      // Attempt to load the panel
```
did you switch to spaces instead of tabs ? (sorry for being that guy :smile:)

tomneedham replied:
Maybe..... ;) Sorted.

PVince81 commented:
Great stuff. Will this be backward compatible ? Basically if apps still provide settings as before, have code that auto-register those through the new settings manager.

tomneedham commented:
> Will this be backward compatible ?

Yes. We can catch the old calls and group them together.

DeepDiver1975 commented:
````
Parse error: ./tests/lib/Settings/ManagerTest.php:43

    41|   public function testLoadPanels() {

    42| 

  > 43|   public function testGetPanels() {}

    44| 

    45|   public function testSortOrder() {}

Unexpected 'public' (T_PUBLIC)

make: *** [test-php-lint] Error 1

Commit:
Handle legacy settings panels

DeepDiver1975 commented on lib/private/Settings/SettingsManager.php:
```diff
@@ -149,8 +151,11 @@ private function getBuiltInPanels() {
             ],
-            'admin' => []
+            'admin' => [
+                'OC\Settings\Panels\Admin\Legacy',
```
Please use 

```php
use OC\Settings\Panels\Admin\Legacy;

....

Legacy::class
```

DeepDiver1975 commented on lib/private/Settings/SettingsManager.php:
```diff
@@ -161,6 +166,8 @@ private function getBuiltInPanels() {
     private function getBuiltInPanel($className) {
         $panels = [
             'OC\Settings\Panels\Personal\Profile' => new \OC\Settings\Panels\Personal\Profile($this->config, $this->groupManager, $this->userSession),
```
is that necessary? the di magic should be capable of instantiating all classes

DeepDiver1975 commented on settings/Controller/SettingsPageController.php:
```diff
@@ -59,8 +60,8 @@ public function __construct($appName,
 	*/
-	public function getPersonal($sectionID='general') {
-		$this->currentSectionID = $sectionID;
+	public function getPersonal($sectionid) {
```
$sectionId

DeepDiver1975 commented on settings/Panels/Admin/Legacy.php:
```diff
@@ -0,0 +1,50 @@
+
+class Legacy implements IPanel {
+
+    public function __construct() {}
```
empty? kill it

DeepDiver1975 commented on lib/private/Settings/Section.php:
```diff
@@ -0,0 +1,51 @@
+ */
+class Section implements ISection {
+
+    protected $id, $name, $priority;
```
each on it's own line please and also add

```php
/** @var int */

DeepDiver1975 commented on lib/public/Settings/ISettingsManager.php:
```diff
@@ -0,0 +1,61 @@
+   * @since 9.2
+   * @return array of OCP\Settings\ISection (sorted by priority)
+   */
+  public function getPersonalSections();
```
indent looks off

DeepDiver1975 commented on settings/Application.php:
```diff
@@ -60,9 +61,25 @@ public function __construct(array $urlParams=[]){
 		 * Controllers
 		 */
+		 $container->registerService('SettingsPageController', function(IContainer $c) {
+ 			return new SettingsPageController(
```
controllers should be loadable via di magic - no need to register them anymore

DeepDiver1975 commented on settings/Controller/SettingsPageController.php:
```diff
@@ -0,0 +1,137 @@
+	 */
+	public function __construct($appName,
+								IRequest $request,
+						        ISettingsManager $settingsManager,
```
indent

DeepDiver1975 commented on settings/Controller/SettingsPageController.php:
```diff
@@ -0,0 +1,137 @@
+		} else if($type == 'admin') {
+			$sections = $this->settingsManager->getAdminSections();
+			$panels = $this->settingsManager->getAdminPanels($this->currentSectionID);
+		}
```
missing else block

Commit:
Move over existing built in panels. Still needs JS work

tomneedham commented:
@felixheidecke Hey! Maybe you can help me decide how we should categorise the existing settings panels in a way that makes most sense for the user?

See the random list I've been working on here: https://github.com/owncloud/core/pull/26449/files#diff-879bc94ed8e636501bd669b9f1d9f17cR147

Basically, legacy uses of registerPersonal and registerAdmin will just go into the 'additional' sections until they get moved to relevant ones.

DeepDiver1975 commented:
Please fix ci - thx

DeepDiver1975 commented on settings/Panels/Admin/BackgroundJobs.php:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    public function getName() {
+        return 'Backgroudn Jobs';
```
typo - do we need translations on this?

DeepDiver1975 commented on settings/templates/panels/personal/profile.php:
```diff
@@ -1,3 +1,8 @@
+<?php
+\OC_Util::addScript('files', 'jquery.fileupload');
```
```php
script('files', 'jquery.fileupload')

DeepDiver1975 commented:
on personal page:

````
personal.js?v=f23f455…:367 Uncaught TypeError: Cannot read property 'AuthTokenCollection' of undefined

DeepDiver1975 commented:
and

```
personal.js?v=f23f455…:288 Uncaught TypeError: $(...).fileupload is not a function

DeepDiver1975 commented:
on admin all section headings are duplicated

DeepDiver1975 committed:
Kill getName on IPanel

DeepDiver1975 committed:
Adding apps mgmt as admin section

Commit:
Add language switcher

Commit:
Delete old personal settings controller

Commit:
Fix sync client urls

Commit:
Fix JS loading for personal settings panels: profile

tomneedham commented:
Not too hot on the authtoken stuff but I think we need a fix for this: https://github.com/owncloud/core/pull/26449/files#diff-a1c52fafc7d1f682c8849c5c62297682L376

We could add a check for this in the controller, find the relevant section given the panel id and then redirect to it?

tomneedham commented:
Also https://github.com/owncloud/core/blob/master/settings/routes.php#L41 is being overwritten by the route below which means the authtoken panel is broken at the moment.

Commit:
Move shipped apps to new admin settings registration

Commit:
Fix section styling on settings panels

tomneedham commented:
- [ ] Fix file sharing settings saving / initial value retrieval

DeepDiver1975 committed:
- fix sort order sections
- adjust some styling
- remote filesexternal section - now provided by the app itself
- code cleanup

DeepDiver1975 committed:
Fix PHPDoc on public interface

felixheidecke commented:
Unifying **personal** and **administrator** under a general **settings** Sectoin is a good idea, since its a *memorized* term in modern UIs. (See mobile phones, desktop software and such).

Moving to *separate* sites is quiet fine, since you usualy do'nt casualy browse the setting but rather look for specific settings.

The only downside is the page reloads but this can be resolved with ajax in the future.

@tomneedham If you need more design / UI input, pls ask.

tomneedham commented:
> Unifying personal and administrator under a general settings section is a good idea, since its a memorized term in modern UIs. (See mobile phones, desktop software and such).

Awesome thanks @felixheidecke. We'll get this one in first which has backend changes then we can work towards that + ajax loading :)

Commit:
Fix sharing settings loading in admin panel

Commit:
Add some padding to help with UI

Commit:
Add active navigation styling

Commit:
Fix navigation to admin and add @deprecated label to old methods

Commit:
Fix documentation link styling on panels and personal sort order

Commit:
Fix logging and cron panels

Commit:
Fix mail and files admin panels

tomneedham commented:
Still to be resolved, some of the existing admin panels used /settings/admin/$x urls to send back data to the server but these clash with the new routing of /settings/admin/$sectionid. Either we moves the URLs to something like /settings/admin/ajax/$whatever and reserve 'ajax' as a section id, or we reserve all of the existing routes that are used so they cannot be used as section ids? Thoughts @DeepDiver1975 ?

Commit:
Add quota to personal panels

Commit:
Migrate encryption and external panels. Change ajax routes. Add error for when no panels exist.

DeepDiver1975 commented on apps/files_external/appinfo/info.xml:
```diff
@@ -39,5 +39,6 @@
 	</commands>
 	<settings>
 		<admin>OCA\Files_External\Panels\Admin</admin>
+		<admin>OCA\Files_External\Panels\Personal</admin>
```
admin? should be personal right?

DeepDiver1975 commented on settings/Panels/Admin/Encryption.php:
```diff
@@ -32,6 +32,23 @@ public function getPriority() {
 
     public function getPanel() {
         $tmpl = new Template('settings', 'panels/admin/encryption');
+		$tmpl->assign('encryptionEnabled', \OC::$server->getEncryptionManager()->isEnabled());
```
indent

DeepDiver1975 commented on settings/routes.php:
```diff
@@ -38,32 +38,32 @@
-		['name' => 'Certificate#removePersonalRootCertificate', 'url' => '/settings/personal/certificate/{certificateIdentifier}', 'verb' => 'DELETE'],
-		['name' => 'Certificate#addSystemRootCertificate', 'url' => '/settings/admin/certificate', 'verb' => 'POST'],
-		['name' => 'Certificate#removeSystemRootCertificate', 'url' => '/settings/admin/certificate/{certificateIdentifier}', 'verb' => 'DELETE'],
+		['name' => 'LogSettings#setLogLevel', 'url' => '/settings/admin/ajax/log/level', 'verb' => 'POST'],
```
what about removing admin/ajax?

so in this case: /settings/log/level

tomneedham replied:
I chose not to because I was worried about breaking routes like this https://github.com/owncloud/core/blob/master/settings/js/certificates.js#L5 which take in the settings type, but if this one is safe I could move it.

Commit:
Fix error message when no panels available

Commit:
Move personal federated sharing panel

Commit:
Remove old call to regsiterPersonal

Commit:
Update route for changing password

Commit:
Revert "Update route for changing password"

This reverts commit c1ef3c1fbf98fc565efd72691dc2ed2ec9f83144.

Commit:
Change routing to be backwards compatible

Commit:
IPanel > ISettings

DeepDiver1975 commented:
nice progress @tomneedham 

Next steps:
- [ ] fix jenkins @tomneedham 
- [ ] add necessary unit test @tomneedham 
- [ ] provide reasonable section for admin and personal @PVince81 @felixheidecke THX
- [ ] decide: merge personal and admin now or additional PR? @PVince81

PVince81 commented:
I assume that will only have one level of sections.
Try and make the number of sections in the left sidebar small.
Within the actual page we can still display multiple sub-sections.

@felixheidecke can you check the sections we have a provide a list of revamped sections ?

@DeepDiver1975 @tomneedham I'd say let's merge the "admin" and "personal" sections now in this PR. It will reduce testing time as we won't need to test twice (once in this PR and once sections merged again)

DeepDiver1975 commented:
> @DeepDiver1975 @tomneedham I'd say let's merge the "admin" and "personal" sections now in this PR. It will reduce testing time as we won't need to test twice (once in this PR and once sections merged again)

Agreed!

tomneedham commented:
![screen shot 2016-12-01 at 12 47 16](https://cloud.githubusercontent.com/assets/660805/20794501/924415be-b7c4-11e6-996a-29e2c133fba5.png)
@PVince81

PVince81 commented:
Hmm, interesting. I didn't think of having "Admin" and "Personal" as main sections, I thought we'd just have all sections directly within the sidebar without differenciation. But maybe your approach is better to keep some structure. So your proposal looks good.

tomneedham committed:
Merge admin and personal settings to one page

tomneedham committed:
Port over admin set upload to new route

tomneedham commented:
> Hmm, interesting. I didn't think of having "Admin" and "Personal" as main sections, I thought we'd just have all sections directly within the sidebar without differenciation

Ah! Funny how two different minds think. This was the easiest option from an implementation point of view, but also keeps it clear for the user.

tomneedham committed:
Load core css last, fix php doc

tomneedham committed:
Fix security warnings panel and merge files_external

tomneedham commented:
![screen shot 2016-12-01 at 17 50 41](https://cloud.githubusercontent.com/assets/660805/20805477/bb6ef678-b7ee-11e6-9a8a-97195a6da0c9.png)

tomneedham committed:
Fix sort order of admin panels

DeepDiver1975 commented on .user.ini:
```diff
@@ -1,5 +1,5 @@
-upload_max_filesize=513M
-post_max_size=513M
+upload_max_filesize=700M
```
please revert this

DeepDiver1975 commented on apps/encryption/lib/Panels/Admin.php:
```diff
@@ -0,0 +1,70 @@
+    }
+}
+
+?>
```
no closing php tag please

DeepDiver1975 commented on apps/encryption/lib/Panels/Personal.php:
```diff
@@ -0,0 +1,64 @@
+    }
+}
+
+?>
```
kill

DeepDiver1975 commented on apps/federatedfilesharing/appinfo/info.xml:
```diff
@@ -11,4 +11,8 @@
     </dependencies>
+    <settings>
+        <admin>OCA\FederatedFileSharing\AdminPanel</admin>
+		<personal>OCA\FederatedFileSharing\PersonalPanel</personal>
```
indent

tomneedham committed:
Revert earlier user.ini alteration

tomneedham committed:
kil end php tags and add phpdoc

tomneedham committed:
Add tests for settingsManager and settingsPageController

tomneedham committed:
Fix indentation

tomneedham commented:
@DeepDiver1975 I'm slightly concerned about my port of the encryption app and how it handles displaying the appropriate settings panels. Can we get someone who is up to speed on that to give this a good test?

tomneedham commented:
- [ ] BUG: Settings not accessible for non admin users: get 403 access denied

PVince81 commented:
Unrelated failure... have seen this on many PRs

```
14:38:32 name: smb-windows
14:39:33 ....................
14:39:33 [ERROR] Server not reachable
14:39:33 stop: stop-smb-windows.sh
```

PVince81 commented:
- [ ] move encryption to its own section

Please move encryption under its own section to make it easier to find. Currently for the personal settings it's under "Security" but under the admin part there is a bit in "General" and another bit in "Security".

I'm not sure if I'd put encryption under security but rather on an own "encryption" section.

PVince81 commented:
- [ ] adjust all section header styles to use the same class

The one of encryption is different:
![sections-style1](https://cloud.githubusercontent.com/assets/277525/20879290/55c729b6-bad3-11e6-9cd4-f9824144f3d3.png)
vs
![sections2](https://cloud.githubusercontent.com/assets/277525/20879294/59f4f112-bad3-11e6-9d13-ccb1b7208a4a.png)

PVince81 commented:
- [ ] search box is hidden: now that one cannot use ctrl+f any more to find in which section a thing is, it would be nice to have a search box. Might be tricky as we'd need an index of all the section contents in advance or at least their heading -> raise ticket

PVince81 commented:
- [ ] Move "Federation" from "Additional settings" into the "Sharing" section ? Or have a dedicated federation section ?

DeepDiver1975 commented:
> search box is hidden: now that one cannot use ctrl+f any more to find in which section a thing is, it would be nice to have a search box. Might be tricky as we'd need an index of all the section contents in advance or at least their heading -> raise ticket

is search still needed with this apporach?
I mean: it can only search client side and the currently loaded page will be rather small compared to the 'old' approach where everything is loaded

PVince81 commented:
> is search still needed with this apporach?

From a user's point of view, yes. Especially now that we revamped the section
Use case: "where do I manage encryption ?". An admin would type "encryption" in the search field and the sections containing encryption would be highlighted. In the past an admin would just "ctrl+f" and type encryption to find it.
Now with the new sections you need to click every section and ctrl+f to search them. So this is a UX regression, but not critical enough to block this PR. Something to thing about in the future.

DeepDiver1975 committed:
Allow non-admins to access settings

tomneedham committed:
Group encryption panels, remove legacy controllers

tomneedham committed:
Migrate federation app to new settings panels

tomneedham committed:
Use correct class on settings panel titles

DeepDiver1975 commented:
please rebase - https://github.com/owncloud/core/pull/26588 was merged

tomneedham commented:
> search box is hidden: now that one cannot use ctrl+f any more to find in which section a thing is, it would be nice to have a search box. Might be tricky as we'd need an index of all the section contents in advance or at least their heading -> raise ticket

Ticket raised: https://github.com/owncloud/core/issues/26765

tomneedham commented:
> adjust all section header styles to use the same class

@PVince81 Could this be a caching issue? I checked these out and they should be consistent.

PVince81 commented:
I keep being confused by the left section, thinking that "Admin" is selected while it's only the header.
Maybe we need a different style ? The header style is too strong compared to the actual selection style.

It seems that in the other app navigations (ex: files left sidebar) we also only use bold for selections, so we should probably keep the selection style but improve the header styles to be less strong.

@felixheidecke any input on that ?

PVince81 commented:
![section-style](https://cloud.githubusercontent.com/assets/277525/20887829/8d96e772-bafc-11e6-82fb-2b3c48ef2c52.png)

or we really change it to not have section headers and merge everything.

This means settings aren't "personal" or "admin" any more, but only the fact that they render or not will change depending on the user role.

So "General" would contain the sections for both admin and personal for an admin, but only personal for a non-admin.

This would solve the header styling issue by getting rid of the headers and might also reduce my own confusion everytime I need to remember whether I need to look at "Security" under "Personal" or "Security" under "Admin".

What do you think ?

PVince81 commented:
Furthermore if we merge "Admin" with "Personal", this would free more vertical space in the sidebar to have more meaningful and finer sections like "Federation", "Encryption". In total we'd save 2 headers + 3 duplicate sections = 5 entries.

PVince81 commented:
> @PVince81 Could this be a caching issue? I checked these out and they should be consistent.

Maybe. Now with the latest version of this PR I can't go to Personal/Encryption, it says "No panels for this section"

tomneedham commented:
> Maybe. Now with the latest version of this PR I can't go to Personal/Encryption, it says "No panels for this section"

This is because the existing panels have annoying if statements to check if they should display something. So the encryption personal panel registers a panel but then doesn't actually output a page (see https://github.com/owncloud/core/blob/master/apps/encryption/settings/settings-personal.php#L65). I added in the error as a last resort catch but we should do better checking in the future to not even show the section if there are no panels with content. 

Also, for the legacy forms, I have to register a panel that catches OC_App::registerSettings|Personal() but this function them might return an empty array causing exactly the same issue of an empty panel with nothing to display - again, can be removed in the future when everything uses the new panel registration stuff.

tomneedham commented:
I like the UX perspective of merging the settings pages into one but i think think might give us complications which would further draw out this PR. For example, we'd need to map existing urls, whereas at the moment the urls / routing is exactly the same.

Also we will probably end up with quite a few duplicate panels until we an go through and combine the controllers, which might mean you end up with 'Federation' title about 3 three times if you are an admin.

Thoughts from @felixheidecke would be helpful here :)

DeepDiver1975 commented:
@PVince81 I suggest to keep the split of personal and admin for now and get this in.
We can work in a followup pr on further enhancements

DeepDiver1975 committed:
In case a section is not found we create a dummy section

PVince81 commented:
Fine by me, but we need to make sure all sections still work as before, like encryption

DeepDiver1975 commented:
> Fine by me, but we need to make sure all sections still work as before, like encryption

@owncloud/qa @owncloud/qa-team please help - THX

SergioBertolinSG commented:
I think headers Personal and Admin will confuse users. They look like another tab. Perhaps if they were totally different, like a title, removing the need of clicking on it, would be nice. 

Why not having a personal section with the current three (general, security and additional) and removing the headers?

I don't see the point in splitting personal in three, specially with those names general and additional.

PVince81 commented:
@SergioBertolinSG thanks, this is indeed what I felt here https://github.com/owncloud/core/pull/26449#issuecomment-264862338

So far we'd like to merge this PR first and continue merging the sections afterwards.

SergioBertolinSG commented:
> This means settings aren't "personal" or "admin" any more, but only the fact that they render or not will change depending on the user role.
> 
> So "General" would contain the sections for both admin and personal for an admin, but only personal for a non-admin.

I find this also confusing, coming from a totally differentiated sections, external storage part for example would be very confusing for an admin.

 
> I don't see the point in splitting personal in three, specially with those names general and additional.
Ok, I answer to myself, there are more sections in personal (external storage, activity, etc) so I see now the point. Btw where are the sessions? I'm not seeing them on settings.

SergioBertolinSG commented:
When a menu item is entered and there is no info to show, it tells the user that an error has occurred, which is not true:

![screen shot 2016-12-07 at 13 58 23](https://cloud.githubusercontent.com/assets/9048796/20968632/8ff12d9a-bc85-11e6-8089-3f52805ecbeb.png)

It should say that there is nothing to show, but no errors.

DeepDiver1975 commented:
> It should say that there is nothing to show, but no errors.

or we simply don't show the section if there is nothing inside

DeepDiver1975 commented:
> Btw where are the sessions? I'm not seeing them on settings.

currently hard coded - will be read from the xml as well

tomneedham commented:
> currently hard coded - will be read from the xml as well

are we talking about sessions or sections here? @SergioBertolinSG @DeepDiver1975 ...?

tomneedham commented:
> or we simply don't show the section if there is nothing inside

Problem we have with this right now is that some forms are deciding if they should output a panel on template runtime, so they register a panel and then decide not to show anything, see https://github.com/owncloud/core/blob/master/apps/encryption/settings/settings-personal.php#L65

Need to have a think if we can evaluate this before we output the sections - which is why I added in the message for the moment.

tomneedham commented:
@SergioBertolinSG Which app specifically do you think is missing at the moment?

tomneedham commented:
> I think headers Personal and Admin will confuse users. They look like another tab. Perhaps if they were totally different, like a title, removing the need of clicking on it, would be nice.

I can appreciate this. This potentially has a much lesser effect if we decide to implement icons as proposed / implemented here: https://github.com/owncloud/core/pull/26759. See screenshot.

SergioBertolinSG commented:
> I can appreciate this. This potentially has a much lesser effect if we decide to implement icons as proposed / implemented here: #26759. See screenshot.

Yes I like this approach.

SergioBertolinSG commented:
> are we talking about sessions or sections here? @SergioBertolinSG @DeepDiver1975 ...?

Sessions, in personal:

<img width="1043" alt="screen shot 2016-12-09 at 10 37 02" src="https://cloud.githubusercontent.com/assets/9048796/21044290/82197ffc-bdfb-11e6-91f0-091e0830e733.png">

DeepDiver1975 commented:
I'll try to squash this to reasonable commits ...

SergioBertolinSG commented:
When going into the item personal/security I am getting an internal server error:

```
{"reqId":"UKUzKvHfbjIYF6xeue+7","remoteAddr":"HIDDEN_IP","app":"index","message":"Exception: {\"Exception\":\"OCP\\\\AppFramework\\\\QueryException\",\"Message\":\"Could not resolve OCA\\\\Files_External\\\\Service\\\\UserStoragesService! Class OCA\\\\Files_External\\\\Service\\\\UserStoragesService does not exist\",\"Code\":0,\"Trace\":\"#0 \\\/opt\\\/owncloud\\\/lib\\\/private\\\/AppFramework\\\/Utility\\\/SimpleContainer.php(113): OC\\\\AppFramework\\\\Utility\\\\SimpleContainer->resolve('OCA\\\\\\\\Files_Exter...')\\n#1 \\\/opt\\\/owncloud\\\/apps\\\/files_external\\\/lib\\\/Panels\\\/Personal.php(24): OC\\\\AppFramework\\\\Utility\\\\SimpleContainer->query('OCA\\\\\\\\Files_Exter...')\\n#2 \\\/opt\\\/owncloud\\\/settings\\\/Controller\\\/SettingsPageController.php(144): OCA\\\\Files_External\\\\Panels\\\\Personal->getPanel()\\n#3 \\\/opt\\\/owncloud\\\/settings\\\/Controller\\\/SettingsPageController.php(88): OC\\\\Settings\\\\Controller\\\\SettingsPageController->getPanelsData(Array)\\n#4 [internal function]: OC\\\\Settings\\\\Controller\\\\SettingsPageController->getPersonal('security')\\n#5 \\\/opt\\\/owncloud\\\/lib\\\/private\\\/AppFramework\\\/Http\\\/Dispatcher.php(159): call_user_func_array(Array, Array)\\n#6 \\\/opt\\\/owncloud\\\/lib\\\/private\\\/AppFramework\\\/Http\\\/Dispatcher.php(89): OC\\\\AppFramework\\\\Http\\\\Dispatcher->executeController(Object(OC\\\\Settings\\\\Controller\\\\SettingsPageController), 'getPersonal')\\n#7 \\\/opt\\\/owncloud\\\/lib\\\/private\\\/AppFramework\\\/App.php(99): OC\\\\AppFramework\\\\Http\\\\Dispatcher->dispatch(Object(OC\\\\Settings\\\\Controller\\\\SettingsPageController), 'getPersonal')\\n#8 \\\/opt\\\/owncloud\\\/lib\\\/private\\\/AppFramework\\\/Routing\\\/RouteActionHandler.php(46): OC\\\\AppFramework\\\\App::main('SettingsPageCon...', 'getPersonal', Object(OC\\\\AppFramework\\\\DependencyInjection\\\\DIContainer), Array)\\n#9 [internal function]: OC\\\\AppFramework\\\\Routing\\\\RouteActionHandler->__invoke(Array)\\n#10 \\\/opt\\\/owncloud\\\/lib\\\/private\\\/Route\\\/Router.php(299): call_user_func(Object(OC\\\\AppFramework\\\\Routing\\\\RouteActionHandler), Array)\\n#11 \\\/opt\\\/owncloud\\\/lib\\\/base.php(894): OC\\\\Route\\\\Router->match('\\\/settings\\\/perso...')\\n#12 \\\/opt\\\/owncloud\\\/index.php(46): OC::handleRequest()\\n#13 {main}\",\"File\":\"\\\/opt\\\/owncloud\\\/lib\\\/private\\\/AppFramework\\\/Utility\\\/SimpleContainer.php\",\"Line\":98}","level":3,"time":"2016-12-09T12:12:22+00:00","method":"GET","url":"\/index.php\/settings\/personal?sectionid=security","user":"admin"}
```

SergioBertolinSG commented:
External storage doesn't appear in personal settings. Also I think it should have it's own section, not be located in admin/general.

SergioBertolinSG commented:
With apps section included in settings, will apps section from the left menu be removed?

SergioBertolinSG commented:
Click on admin/additional returns a forbidden answer to a GET request:

Request URL: HIDDEN_HOST/data/htaccesstest.txt?t=1481288588517
Request Method:GET
Status Code:403 Forbidden

SergioBertolinSG commented:
I think well known apps (At least those which are included in the tarball) should have their own section. For example LDAP. Currently it appears in additional.

tomneedham commented:
> I think well known apps (At least those which are included in the tarball) should have their own section. For example LDAP. Currently it appears in additional.

I think this is a design decision. I was under the impression that 'Sections' (on the left) weren't for apps to just register their own, but to group settings panels into sensible groups based on their topic. But I do agree we have room to expand these to make use of this feature. 

I think 'Storage' would be a sensible section. 'Users' (for LDAP and user management) would make sense if users was under this admin structure.

PVince81 commented:
> I was under the impression that 'Sections' (on the left) weren't for apps to just register their own

I'd also have thought so initially but now that we walked the path it seems that it's better if apps can register their own sections or add themselves to existing sections by id.

tomneedham commented:
> I'd also have thought so initially but now that we walked the path it seems that it's better if apps can register their own sections or add themselves to existing sections by id.

Yeah exactly :) We're on the same page here. Apps adding their own sections will come in another PR. 

I'll add 'Storage' and stick the storage stuff in there as that seems a pretty common one. 

I'd say we focus on getting bugs fixed in this PR so we can get it in and then tweak the exact wording.

tomneedham committed:
Add in browser sessions and rename apppassword panel to tokens

tomneedham commented:
<img width="1074" alt="screen shot 2016-12-09 at 19 55 43" src="https://cloud.githubusercontent.com/assets/660805/21062614/99c75d56-be49-11e6-9651-c6e602ec94e7.png">
@SergioBertolinSG Fixed, the html didn't make it over! Good spot.

tomneedham committed:
Add storage section and fix files_external personal panel

tomneedham commented:
TODO: add unit tests for the upgraded panels:

 - [ ] OCA\Encryption\Panels\Admin
 - [ ] OCA\Encryption\Panels\Personal
 - [ ] OCA\FederatedFileSharing\AdminPanel
 - [ ] OCA\FederatedFileSharing\PersonalPanel
 - [ ] OCA\Federation\Panels\Admin
 - [ ] OCA\Files\Panels\Admin
 - [ ] OCA\Files_External\Panels\Admin
 - [ ] OCA\Files_External\Panels\Admin
 - [ ] OCA\UpdateNotification\AdminController
 - [ ] OC\Settings\Panels\Admin\Apps
 - [ ] OC\Settings\Panels\Admin\BackgroundJobs
 - [ ] OC\Settings\Panels\Admin\Certificates
 - [ ] OC\Settings\Panels\Admin\Encryption
 - [ ] OC\Settings\Panels\Admin\FileSharing
 - [ ] OC\Settings\Panels\Admin\Legacy
 - [ ] OC\Settings\Panels\Admin\Logging
 - [ ] OC\Settings\Panels\Admin\Mail
 - [ ] OC\Settings\Panels\Admin\SecurityWarning
 - [ ] OC\Settings\Panels\Admin\Tips
 - [ ] OC\Settings\Panels\Admin\Updater
 - [ ] OC\Settings\Panels\Legacy

tomneedham committed:
Fix error message when error retrieving panel

tomneedham committed:
Remove old admin settings controller

tomneedham committed:
Add framework of panel tests

tomneedham committed:
More tests for settings panels

tomneedham committed:
Add encryption app panel tests

tomneedham committed:
Add unit tests for remaining panels

tomneedham committed:
Add federation panel tests

tomneedham committed:
Add federatedfilesharing admin panel test template

tomneedham committed:
Add test for admin apps panel

tomneedham committed:
Add DI for federatedfilesharing panel and add unit test

tomneedham committed:
Fix DI for settings manager tests

tomneedham committed:
Add getTransationIsolation to IDBConnection

cdamken commented:
Does encryption still appears even if it's disabled?

tomneedham commented:
@cdamken Yes there is a built in panel (https://github.com/owncloud/core/pull/26449/files#diff-e3dd6e10696542d6118a8357ae61e20c) which is always loaded allowing you to enable it. But please help test :)

PVince81 commented:
@tomneedham anything left to do or is this ready for re-review ?

tomneedham commented:
@PVince81 Ready for review please. I spoke to Thomas last week about the the failing test complaining about the Mock object not matching the interface - maybe you can take a look?

tomneedham committed:
App supplied sections should appear just above additional panels in the list

tomneedham commented:
Just added a fix to push app-defined sections to the bottom of the list

DeepDiver1975 commented:
````
PHP Fatal error:  Class OC\AppFramework\Db\Db contains 1 abstract method and must therefore be declared abstract or implement the remaining methods (OCP\IDBConnection::getTransactionIsolation) in /var/lib/jenkins/workspace/owncloud-core_core_PR-26449-ZMAJD5DRZL7TNCVFUCOZGMUDLC37Y6X2F73FIL4ULXIPIZYCJYIA/lib/private/AppFramework/Db/Db.php on line 38
10:17:19 Makefile:147: recipe for target 'test-php' failed

DeepDiver1975 committed:
Add missing method getTransactionIsolation

DeepDiver1975 committed:
Properly mock IFactory

DeepDiver1975 commented:
@tomneedham please add the fix as of  	ee0b576 to ProfileTest

tomneedham committed:
Fix user profile panel tests

tomneedham committed:
Fix broken profiles panel and settingsPageController tests

DeepDiver1975 commented:
some more broken tests ...

please run unit tests locally - thx

````
 OCA\Files_External\Tests\Panels\AdminTest.testGetSection
 Stack Trace

OCA\Files_External\Tests\Panels\AdminTest::testGetSection
TypeError: Argument 1 passed to OCA\Files_External\Panels\Admin::__construct() must implement interface OCP\Files\External\Service\IGlobalStoragesService, instance of Mock_IStoragesBackendService_acbe3e89 given, called in /var/lib/jenkins/workspace/owncloud-core_core_PR-26449-ZMAJD5DRZL7TNCVFUCOZGMUDLC37Y6X2F73FIL4ULXIPIZYCJYIA/apps/files_external/tests/Panels/AdminTest.php on line 40

/var/lib/jenkins/workspace/owncloud-core_core_PR-26449-ZMAJD5DRZL7TNCVFUCOZGMUDLC37Y6X2F73FIL4ULXIPIZYCJYIA/apps/files_external/lib/Panels/Admin.php:22
/var/lib/jenkins/workspace/owncloud-core_core_PR-26449-ZMAJD5DRZL7TNCVFUCOZGMUDLC37Y6X2F73FIL4ULXIPIZYCJYIA/apps/files_external/tests/Panels/AdminTest.php:40

tomneedham committed:
Allow BackgroundJobsTest to use legacy template functions with db

tomneedham committed:
Fix files_external settings panel tests DI issues

tomneedham commented:
🎉

DeepDiver1975 commented on apps/encryption/lib/Panels/Personal.php:
```diff
@@ -0,0 +1,61 @@
+				\OC::$server->getConfig(),
+				\OC::$server->getUserManager());
+
+		$keyManager = new \OCA\Encryption\KeyManager(
```
unused

DeepDiver1975 commented on apps/files_external/lib/Panels/Admin.php:
```diff
@@ -0,0 +1,35 @@
+
+    public function getPanel() {
+		// we must use the same container
+		$appContainer = \OC_Mount_Config::$app->getContainer();
```
unused

DeepDiver1975 commented on apps/files_external/lib/Panels/Personal.php:
```diff
@@ -0,0 +1,35 @@
+
+    public function getPanel() {
+		// we must use the same container
+		$appContainer = \OC_Mount_Config::$app->getContainer();
```
unused

DeepDiver1975 commented on settings/Panels/Admin/FileSharing.php:
```diff
@@ -0,0 +1,59 @@
+		$template->assign('onlyShareWithGroupMembers', \OC\Share\Share::shareWithGroupMembersOnly());
+		$template->assign('allowMailNotification', $config->getAppValue('core', 'shareapi_allow_mail_notification', 'no'));
+		$template->assign('allowShareDialogUserEnumeration', $config->getAppValue('core', 'shareapi_allow_share_dialog_user_enumeration', 'yes'));
+		$template->assign('shareExcludeGroups', $excludeGroups);
```
excludeGroups is undefined

DeepDiver1975 commented on settings/Panels/Personal/Clients.php:
```diff
@@ -0,0 +1,62 @@
+
+class Clients implements ISettings {
+
+    /** @var OCP\IConfig */
```
missing \

DeepDiver1975 commented on settings/Panels/Personal/Clients.php:
```diff
@@ -0,0 +1,62 @@
+
+    public function __construct(IConfig $config, Defaults $defaults) {
+        $this->config = $config;
+        $this->defaults = $defaults;
```
$this->default declaration is missing - please review all other classes

DeepDiver1975 commented on settings/Panels/Personal/Profile.php:
```diff
@@ -0,0 +1,128 @@
+
+class Profile implements ISettings {
+
+    /* @var Config */
```
Config is unknown

DeepDiver1975 commented on settings/Controller/SettingsPageController.php:
```diff
@@ -0,0 +1,155 @@
+	 * @param string $sectionid
+	 * @return \OCP\AppFramework\Http\TemplateResponse
+	 */
+	public function getPersonal($sectionid='general') {
```
$sectionId

tomneedham committed:
Code style fixes

davitol reviewed (changes requested):
* Users cannot change their Passwords


* 'En' instead of 'English' in dropdown menu (same in other languages)

![screen shot 2017-02-03 at 13 22 58](https://cloud.githubusercontent.com/assets/10596560/22591266/eedc18d6-ea13-11e6-81b7-ae6d293de8ee.png)




* 'Test email settings' button 'fakes' result. Green when no e-mail was set

![screen shot 2017-02-03 at 13 41 37](https://cloud.githubusercontent.com/assets/10596560/22591831/d91f42e0-ea16-11e6-93d1-0ef71ccf140a.png)
* Left side Personal Section is useless
![screen shot 2017-02-03 at 13 47 27](https://cloud.githubusercontent.com/assets/10596560/22592028/1055b4be-ea18-11e6-8e97-f2e589d039cf.png)

and create these logs when clicking on it:

```
{"reqId":"h8TWP1aBEEEH57NFFHrk","remoteAddr":"82.159.139.58","app":"PHP","message":"Illegal offset type in isset or empty at \/opt\/owncloud\/lib\/private\/Cache\/CappedMemoryCache.php#41","level":3,"time":"2017-02-03T13:10:16+00:00","method":"GET","url":"\/index.php\/settings\/personal?sectionid=general","user":"admin"}
{"reqId":"h8TWP1aBEEEH57NFFHrk","remoteAddr":"82.159.139.58","app":"PHP","message":"Object of class OC\\User\\User could not be converted to string at \/opt\/owncloud\/lib\/composer\/doctrine\/dbal\/lib\/Doctrine\/DBAL\/Driver\/PDOStatement.php#91","level":3,"time":"2017-02-03T13:10:16+00:00","method":"GET","url":"\/index.php\/settings\/personal?sectionid=general","user":"admin"}
{"reqId":"h8TWP1aBEEEH57NFFHrk","remoteAddr":"82.159.139.58","app":"PHP","message":"Illegal offset type at \/opt\/owncloud\/lib\/private\/Cache\/CappedMemoryCache.php#49","level":3,"time":"2017-02-03T13:10:16+00:00","method":"GET","url":"\/index.php\/settings\/personal?sectionid=general","user":"admin"}
```

* SideBar for Admin menu; Only general section works

![screen shot 2017-02-03 at 14 17 58](https://cloud.githubusercontent.com/assets/10596560/22592853/f2334eca-ea1b-11e6-80bf-3d58edd26a1c.png)

 
![screen shot 2017-02-03 at 14 47 15](https://cloud.githubusercontent.com/assets/10596560/22593652/e5d5e062-ea1f-11e6-9307-56c931e3ca52.png)

tomneedham committed:
Fix route parameter name to match url

tomneedham commented:
Thanks @davitol for checking this. In one of my later commits I changed a variable name and it killed the routing for anything other than 'general' settings. Fixed that now, and the route for the setting the password.

davitol commented:
@tomneedham 

* Clicking on setting menu, it spits the following logs in owncloud.log

```
{"reqId":"ofquu\/bHI\/mXaOSjJly+","remoteAddr":"82.159.139.58","app":"PHP","message":"Illegal offset type in isset or empty at \/opt\/owncloud\/lib\/private\/Cache\/CappedMemoryCache.php#41","level":3,"time":"2017-02-07T09:13:27+00:00","method":"GET","url":"\/index.php\/settings\/personal","user":"admin"}
{"reqId":"ofquu\/bHI\/mXaOSjJly+","remoteAddr":"82.159.139.58","app":"PHP","message":"Object of class OC\\User\\User could not be converted to string at \/opt\/owncloud\/lib\/composer\/doctrine\/dbal\/lib\/Doctrine\/DBAL\/Driver\/PDOStatement.php#91","level":3,"time":"2017-02-07T09:13:27+00:00","method":"GET","url":"\/index.php\/settings\/personal","user":"admin"}
{"reqId":"ofquu\/bHI\/mXaOSjJly+","remoteAddr":"82.159.139.58","app":"PHP","message":"Illegal offset type at \/opt\/owncloud\/lib\/private\/Cache\/CappedMemoryCache.php#49","level":3,"time":"2017-02-07T09:13:27+00:00","method":"GET","url":"\/index.php\/settings\/personal","user":"admin"}
```


* 'En' instead of 'English' in dropdown menu (same in other languages)

* Clicking on additional menu:

![screen shot 2017-02-07 at 10 19 17](https://cloud.githubusercontent.com/assets/10596560/22685072/2a949070-ed1f-11e6-98b8-258127de1291.png)

* IMHO if apps are going to be managed from the settings menu, the apps+ button should be removed from the left side

![screen shot 2017-02-07 at 10 22 52](https://cloud.githubusercontent.com/assets/10596560/22685166/97217c6c-ed1f-11e6-97e5-20317afdec0a.png)

* The following error keeps showing when clicking on any admin section

![screen shot 2017-02-03 at 14 47 15](https://cloud.githubusercontent.com/assets/10596560/22685424/88feb748-ed20-11e6-8a31-6b704b0dc8d6.png)

davitol commented:
@tomneedham 



> Thanks @davitol for checking this. In one of my later commits I changed a variable name and it killed the routing for anything other than 'general' settings. Fixed that now, and the route for the setting the password.

:+1: It is fixed



> Can you please verify the email functionality against master? The route is being fired correctly and this PR doesn't touch the logic here so bit bewildered as to why it would change, but if its not a problem on master I'll dig into it.

Now it seems to work fine, cannot reproduce it again, so maybe is a corner case not related with this PR

DeepDiver1975 commented:
> IMHO if apps are going to be managed from the settings menu, the apps+ button should be removed from the left side

this is part of another pr: https://github.com/owncloud/core/pull/26763

davitol commented:
Logs are not shown via webUI:

This PR:

![screen shot 2017-02-07 at 13 23 36](https://cloud.githubusercontent.com/assets/10596560/22691122/38f222c6-ed39-11e6-8086-324f4eb738c3.png)

Previous versions:
![screen shot 2017-02-07 at 13 22 49](https://cloud.githubusercontent.com/assets/10596560/22691152/5da9051c-ed39-11e6-9da1-a3da7d3450a1.png)

Note: TypeError: 'OC.Log is undefined' is spotted in Firebug when we change 'what to log'
